google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/apis/DependencyVersionWithBomInspection.java

```diff
@@ -0,0 +1,172 @@
+
+    checkDependencyVersionWithBom(projectModel, holder);
+  }
+
+  /**
+   * Locates google-cloud-java dependencies that have a version tag defined when a BOM is also
+   * imported in the pom.xml. Displays an inspection warning and suggests a quickfix to delete the
+   * version from the dependency.
+   */
+  private void checkDependencyVersionWithBom(
+      MavenDomProjectModel projectModel, DomElementAnnotationHolder holder) {
+    Module module = projectModel.getModule();
+
+    if (module == null) {
+      return;
+    }
+
+    Set<CloudLibrary> cloudLibraries =
+        CloudLibraryProjectState.getInstance(module.getProject()).getCloudLibraries(module);
+
+    if (cloudLibraries.isEmpty()) {
+      return;
+    }
+
+    if (!CloudLibraryProjectState.getInstance(module.getProject())
+        .getCloudLibraryBom(module)
+        .isPresent()) {
+      return;
+    }
+
+    projectModel
+        .getDependencies()
+        .getDependencies()
+        .forEach(
+            dependency -> {
+              if (hasVersion(dependency) && isCloudLibraryDependency(dependency, cloudLibraries)) {
+                holder.createProblem(
+                    dependency.getVersion(),
+                    HighlightSeverity.GENERIC_SERVER_ERROR_OR_WARNING,
+                    GctBundle.message(
+                        "cloud.libraries.version.with.bom.inspection.problem.description"),
+                    new StripDependencyVersionQuickFix());
+              }
+            });
+  }
+
+  /** Checks to see if the {@link MavenDomDependency} has a version defined. */
+  private boolean hasVersion(MavenDomDependency dependency) {
+    return dependency.getVersion().exists();
+  }
+
+  /**
+   * Checks the supplied {@link MavenDomDependency} to see if it is contained in the known set of
+   * managed Google {@link CloudLibrary cloudLibraries}.
+   *
+   * @param dependency the maven dependency we are checking
+   * @param cloudLibraries the set of {@link CloudLibrary cloudLibraries} configured in the project
+   * @return {@code true} if the maven dependency is a google cloud library, and {@code false}
+   *     otherwise
+   */
+  private boolean isCloudLibraryDependency(
+      MavenDomDependency dependency, Set<CloudLibrary> cloudLibraries) {
+    return cloudLibraries
+        .stream()
+        .anyMatch(
+            library ->
+                CloudLibraryUtils.getFirstJavaClientMavenCoordinates(library)
+                    .map(
+                        coords ->
+                            domValueEquals(dependency.getGroupId(), coords.getGroupId())
+                                && domValueEquals(
+                                    dependency.getArtifactId(), coords.getArtifactId()))
+                    .orElse(false));
+  }
+
+  private boolean domValueEquals(GenericDomValue domValue, @Nullable String value) {
+    return domValue.getStringValue() != null && domValue.getStringValue().equals(value);
```

Comment:
Could `domValue` itself be null, worth checking too maybe?

Reply:
I don't think it can (this is the identical situation as the other NPE question you had). But I'll add a check to be safe.